apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts

```diff
@@ -0,0 +1,1541 @@
+                    },
+                  },
+                  {
+                    role: 'Project Manager',
+                    inactiveSinceDate: null,
+                    linkedFrom: {
+                      usersCollection: {
+                        items: [{ email: 'pm4@example.com' }],
+                      },
+                    },
+                  },
+                ],
+              },
+            },
+          },
+        ],
+      };
+      contentfulGraphqlClientMock.request.mockResolvedValueOnce({
+        researchOutputs,
+      });
+      const result = await researchOutputDataProvider.fetchById('output-id');
+
+      expect(result!.contactEmails).toEqual([
+        'pm1@example.com',
+        'pm2@example.com',
+        'pm4@example.com',
+      ]);
+    });
+  });
+  describe('create', () => {
+    beforeEach(() => {
+      environmentMock.createEntry.mockResolvedValue({
+        sys: { id: '1' },
+        publish: jest.fn(),
+      } as unknown as Entry);
+    });
+
+    test('can create a draft output', async () => {
+      const publish = jest.fn();
+      environmentMock.createEntry.mockResolvedValue({
+        sys: { id: '1' },
+        publish,
+      } as unknown as Entry);
+      await researchOutputDataProvider.create(
+        getResearchOutputCreateDataObject(),
+        { publish: false },
+      );
+      expect(environmentMock.createEntry).toHaveBeenCalled();
+      expect(environmentMock.createEntry).toHaveBeenCalledWith(
+        'researchOutputs',
+        expect.anything(),
+      );
+      expect(publish).not.toHaveBeenCalled();
+    });
+
+    test('can create a published output', async () => {
+      const publish = jest.fn();
+      environmentMock.createEntry.mockResolvedValue({
+        sys: { id: '1' },
+        publish,
+      } as unknown as Entry);
+      await researchOutputDataProvider.create(
+        getResearchOutputCreateDataObject(),
+        { publish: true },
+      );
+      expect(environmentMock.createEntry).toHaveBeenCalled();
+      expect(environmentMock.createEntry).toHaveBeenCalledWith(
+        'researchOutputs',
+        expect.anything(),
+      );
+      expect(publish).toHaveBeenCalled();
+    });
+
+    test('returns the id of the created output', async () => {
+      environmentMock.createEntry.mockResolvedValue({
+        sys: { id: 'output-1' },
+      } as unknown as Entry);
+      const result = await researchOutputDataProvider.create(
+        getResearchOutputCreateDataObject(),
+        { publish: false },
+      );
+      expect(result).toEqual('output-1');
+    });
+
+    test('sets external authors ids to contentful references', async () => {
+      await researchOutputDataProvider.create(
+        {
+          ...getResearchOutputCreateDataObject(),
+          authors: [{ userId: 'user-1' }, { externalAuthorId: 'ea-1' }],
+        },
+        { publish: false },
+      );
+      const args = environmentMock.createEntry.mock.lastCall?.[1];
```

Comment:
no way to use  .toBeCalledWith instead?

Reply:
No, it was the several layers of nested objects, which meant a single use of `expect.objectContaining` wasn't sufficient, but replacing with several instances inside one another was fine.